in-progress/6973-refactor-base-contract-interaction.md

```diff
@@ -0,0 +1,1092 @@
+
+
+## Interface
+
+The key operations we want to perform are:
+- Simulating/Reading
+- Proving
+- Submitting
+
+These operations should be accessible from a `Wallet`, which will generally be `Schnorr` or `Signerless`. 
+
+This is analogous to `viem` having a `walletClient` and a `publicClient`.
+
+The main idea in this design is to have the BaseWallet define all the functionality to perform the key operations; the specific entrypoint in use will be able to adapt the `TxExecutionRequest`s produced to its needs.
+
+To accomplish this, we have the BaseWallet construct a `TxExecutionRequestBuilder` that can be passed to various `TxExecutionRequestAdapters`, such as one for the account entrypoint, one for the fee payment method, etc.
+
+Further, we expand `TxExecutionRequest` to allow passing things that we were previously setting directly on the PXE, so simulations can be idempotent.
+
+The interfaces are as follows:
+
+```ts
+
+// new
+export interface ArtifactAndInstance {
+  artifact: ContractArtifact;
+  instance: ContractInstanceWithAddress;
+}
+
+// modified
+export class TxExecutionRequest {
+  constructor(
+    // All these are the same:
+    public origin: AztecAddress,
+    public functionSelector: FunctionSelector,
+    public firstCallArgsHash: Fr,
+    public txContext: TxContext,
+    public argsOfCalls: PackedValues[],
+    public authWitnesses: AuthWitness[],
+    // Add:
+    /**
+     * Transient capsules needed for this execution.
+     */
+    public capsules: Fr[][],
+    /**
+     * Transient contracts needed for this execution.
+     */
+    public transientContracts: ArtifactAndInstance[],
+  ) {}
+  // ...
+}
+
+// new
+export interface InstanceDeploymentParams {
+  constructorName: string;
+  constructorArgs: any;
+  salt: Fr;
+  publicKeysHash: Fr;
+  deployer: AztecAddress;
+}
+
+// new
+export interface DeploymentOptions {
+  registerClass?: boolean;
+  publicDeploy?: boolean;
+}
+
+export interface UserFunctionCall {
+  contractInstance: ContractInstanceWithAddress;
+  functionName: string;
+  args: any;
+  deploymentOptions?: DeploymentOptions;
+  contractArtifact?: ContractArtifact;
+  functionAbi?: FunctionAbi;
+}
+
+// new
+export interface UserRequest {
+  calls: UserFunctionCall[];
+  gasSettings?: GasSettings;
+  paymentMethod?: FeePaymentMethod;
+  from?: AztecAddress;
+  simulatePublicFunctions?: boolean;
+  executionResult?: ExecutionResult; // the raw output of a simulation that can be proven
+  tx?: Tx; // a proven tx to send
+}
+
+export type TxExecutionRequestAdapter = (builder: TxExecutionRequestBuilder, userRequest: UserRequest) => Promise<void>;
+
+export interface TxExecutionRequestComponent {
+  adaptTxExecutionRequest: TxExecutionRequestAdapter;
+}
+
+export interface FeePaymentMethod extends TxExecutionRequestComponent {
+  getEquivalentAztBalance(): Promise<Fr>;
+}
+
+export interface SimulationOutput {
+  tx: Tx;
+  result: DecodedReturn | [];
```

Comment:
This one seem slightly odd to me, you mean `DecodedReturn[]`?

Reply:
Actually, it was an effort to be able to return "no value". But I just wasn't thinking: `[]` is a valid `DecodedReturn`, as is `{}`.